src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,916 @@
+  const size_t n_cbin = (Magnitude.n_cols + downscale - 1) / downscale;
+
+  double o_range, o;
+  o_range = arma::datum::pi / numOrient;
+
+  HistArr = CubeType(n_rbin, n_cbin, numOrient);
+  HistArr.zeros();
+
+  size_t r, c, o1, o2;
+  for(size_t i = 0; i < Magnitude.n_rows; ++i)
+  {
+    for(size_t j = 0; j < Magnitude.n_cols; ++j)
+    {
+      r = i / downscale;
+      c = j / downscale;
+
+      if( interp != 0)
+      {
+        o = Orientation(i, j) / o_range;
+        o1 = ((size_t) o) % numOrient;
+        o2 = (o1 + 1) % numOrient;
+        HistArr(r, c, o1) += Magnitude(i, j) * (1 + (int)o - o);
+        HistArr(r, c, o2) += Magnitude(i, j) * (o - (int) o);
+      }
+      else
+      {
+        o1 = (size_t) (Orientation(i, j) / o_range + 0.5) % numOrient;
+        HistArr(r, c, o1) += Magnitude(i, j);
+      }
+    }
+  }
+
+  HistArr = HistArr / downscale;
+  HistArr = arma::square(HistArr);
+}
+
+/**
+ * Shrink the size of Image by shrink size.
+ * Change color space of Image.
+ * Extract candidate features.
+ * @param InImage, Input Image.
+ * @param regCh, 
+ * @param ssCh, 
+ */
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+GetShrunkChannels(const CubeType& InImage, CubeType& reg_ch,
+                  CubeType& ss_ch, const arma::vec& table)
+{
+  CubeType luv;
+  this->RGB2LUV(InImage, luv, table);
+  const size_t shrink = this->params.Shrink();
+  const size_t grdSmoothRad = this->params.GrdSmoothRad();
+  const size_t numChannels = 13;
+  const size_t rsize = luv.n_rows / shrink;
+  const size_t csize = luv.n_cols / shrink;
+ 
+  CubeType channels(rsize, csize, numChannels);
+
+  size_t slice_idx;
+  for( slice_idx = 0; slice_idx < luv.n_slices; ++slice_idx)
+    this->BilinearInterpolation(luv.slice(slice_idx), rsize, csize,
+                                channels.slice(slice_idx));
+  double scale = 1.0;
+  
+  while(scale >= 0.5)
+  {
+    CubeType Img( (luv.n_rows * scale),
+                   (luv.n_cols * scale),
+                   luv.n_slices );
+
+    for( slice_idx = 0; slice_idx < luv.n_slices; ++slice_idx)
+    {
+      this->BilinearInterpolation(luv.slice(slice_idx), 
+            (luv.n_rows * scale), (luv.n_cols * scale), 
+            Img.slice(slice_idx));
+    }
+    this->ConvTriangle(Img, grdSmoothRad); 
+    MatType Magnitude(InImage.n_rows, InImage.n_cols),
+             Orientation(InImage.n_rows, InImage.n_cols);
+    
+    this->Gradient(Img, Magnitude, Orientation);
+    size_t downscale = std::max(1, (int)(shrink * scale));
+    
+    CubeType Hist;
+    this->Histogram(Magnitude, Orientation,
+                    downscale, 0, Hist);
+    BilinearInterpolation( Magnitude, rsize, csize, channels.slice(slice_idx));
+    slice_idx++;
+    for(size_t i = 0; i < InImage.n_slices; ++i)
+      BilinearInterpolation( Magnitude, rsize, csize,
+                      channels.slice(i + slice_idx));    
+    slice_idx += 3;
```

Comment:
I would recommend to use slice_idx += n_slicesinstead of slice_idx += 3; there is no guarantee that we call the function with a 3 channel image.